NDCube.squeeze method

mihailbankov opened the pull request:
Creates the NDCube.squeeze method as described in the issue, similar to the numpy equivalent.
This method removes (specified) axis with length 1.
Testing for the method is also included.
Resolves #616

DanRyanIrish reviewed (changes requested):
Hi @mihailbankov.  Thanks so much for this PR!  I've suggested a slight refactor to your implementation that doesn't rely on for loops, which should make the code simpler and more efficient.

Also, I suggest that the an error is raised if a non-length-1 axes tries to be squeezed.  This is consistent with `np.squeeze` which many users will already be familiar with.  That change is included in my suggested refactor.

Finally, scalar `NDCube`s aren't supported because WCS's need non-scalar arrays for their transformations between pixel and world values to make sense.  So I've also included an error which is raised if the user tries to squeeze all axes.

Let me know your thoughts.  Thanks again!

mbankovm commented on ndcube/ndcube.py:
```diff
@@ -1186,3 +1186,46 @@ def my_propagate(uncertainty, data, mask, **kwargs):
+        : `~ndcube.NDCube`
+            A new instance with the removed axes.
+        """
+        # List of positions for dimensions to be removed.
+        to_modify = []
+
+        # Checks which dimensions are to be removed.
+        for i, axis_length in enumerate(self.data.shape):
+
+            # Case in which no axis is specified.
+            if axis_length == 1 and axis is None:
+                to_modify.append(i)
+
+            # Case in which axis specified.
+            elif axis_length == 1 and axis is not None and i in axis:
+                to_modify.append(i)
+
+        # Create "mask" for the slicing.
+        arr = np.full(len(self.dimensions),slice(None))
+
+        # The "mask" contains slice(None) for the dimensions we ignore
+        # and 0 for the dimensions that are to be trimmed.
+        for i in to_modify:
+            arr[i] = 0
+
+        # We do the actual slicing.
+        return self[tuple(arr)]
```
Thank you so much for that! I will definitely implement those changes.
I would also like to ask about the test coverage - in the diff it says that the code is not covered by tests, which I find weird since locally the tests run fine and cover everything.

mihailbankov replied:
Thank you so much for that! I will definitely implement those changes.
I would also like to ask about the test coverage - in the diff it says that the code is not covered by tests, which I find weird since locally the tests run fine and cover everything.

DanRyanIrish commented:
At first glance, the test coverage does seem a bit weird.  I would suggest making the changes to the main code, which will also require some changes to the tests.  If the coverage still looks weird, we can take a closer look then.

mihailbankov commented:
I changed the line for isinstance where I put int; On the last line i changed arr to item; Aside from that I think there are no real changes to the code you proposed. The tests were also changed to test the new code.

nabobalis commented on ndcube/ndcube.py:
```diff
@@ -1197,35 +1197,31 @@ def squeeze(self, axis=None):
+            item[shape == 1] = 0
+        else:
+            # For simplicity sake, if axis is scalar make it a tuple.
+            if isinstance(axis, int):
```
Fixed.

DanRyanIrish commented on ndcube/tests/test_ndcube.py:
```diff
@@ -1138,6 +1138,21 @@ def test_to_dask(ndcube_2d_dask):
+    assert np.array_equal(ndcube_4d_ln_l_t_lt[:,:,0,:].dimensions, ndcube_4d_ln_l_t_lt[:,:,0:1,:].squeeze().dimensions)
+    assert np.array_equal(ndcube_4d_ln_l_t_lt[:,:,0,:].dimensions, ndcube_4d_ln_l_t_lt[:,:,0:1,:].squeeze(2).dimensions)
+    assert np.array_equal(ndcube_4d_ln_l_t_lt[:,0,0,:].dimensions, ndcube_4d_ln_l_t_lt[:,0:1,0:1,:].squeeze([1,2]).dimensions)
+
```
```suggestion
assert np.array_equal(ndcube_4d_ln_l_t_lt[:,0:1,0,:].dimensions, ndcube_4d_ln_l_t_lt[:,0:1,0:1,:].squeeze(2).dimensions)

```

DanRyanIrish reviewed (changes requested):
We also need to test whether `.squeeze()` only squeezes the requested axes, rather than all length-1 axes.

DanRyanIrish commented on ndcube/tests/test_ndcube.py:
```diff
@@ -1138,6 +1138,22 @@ def test_to_dask(ndcube_2d_dask):
+    assert np.array_equal(ndcube_4d_ln_l_t_lt[:,:,0,:].dimensions, ndcube_4d_ln_l_t_lt[:,:,0:1,:].squeeze().dimensions)
+    assert np.array_equal(ndcube_4d_ln_l_t_lt[:,:,0,:].dimensions, ndcube_4d_ln_l_t_lt[:,:,0:1,:].squeeze(2).dimensions)
+    assert np.array_equal(ndcube_4d_ln_l_t_lt[:,0,0,:].dimensions, ndcube_4d_ln_l_t_lt[:,0:1,0:1,:].squeeze([1,2]).dimensions)
+assert np.array_equal(ndcube_4d_ln_l_t_lt[:,0:1,0,:].dimensions, ndcube_4d_ln_l_t_lt[:,0:1,0:1,:].squeeze(2).dimensions)
```
```suggestion
    assert np.array_equal(ndcube_4d_ln_l_t_lt[:,0:1,0,:].dimensions, ndcube_4d_ln_l_t_lt[:,0:1,0:1,:].squeeze(2).dimensions)
```

nabobalis commented:
Thanks for the PR @mihailbankov
